### PR TITLE
Bingo

### PR DESCRIPTION
**Pull request by Alchav**
Bingo game, uses a game slot but is run automatically by the server. Bingo calls are shuffled into the item pool. Each Bingo call exists on two Bingo cards. When a line is completed on the Bingo card, a check is sent. This has the effect of backloading item distribution, because you might get many Bingo calls before getting anything back, but later on, a single Bingo call may trigger as many as 6 checks.

### CODE REVIEW

File: WebHostLib/tracker.py [hunk 1]
```diff
@@ -900,6 +900,42 @@ def __renderSuperMetroidTracker(multisave: Dict[str, Any], room: Room, locations
...
 
+
+def __renderBingoTracker(multisave: Dict[str, Any], room: Room, locations: Dict[int, Dict[int, Tuple[int, int, int]]],
+                                inventory: Counter, team: int, player: int, playerName: str,
+                                seed_checks_in_area: Dict[int, Dict[str, int]], checks_done: Dict[str, int], slot_data: Dict) -> str:
+
+    from worlds.bingo.Items import item_table
+    ordered_items = multisave.get('received_items', {}).get((team, player, True), [])
+    # web design is my passion
+    text = "<HTML><HEAD><title>Bingo tracker</title><meta http-equiv=\"refresh\" content=\"15\" />"
```
Comment (black-sliver) [inline WebHostLib/tracker.py:911]: this should probably go in a (jinja) template

File: WebHostLib/tracker.py [hunk 2]
```diff
@@ -900,6 +900,42 @@ def __renderSuperMetroidTracker(multisave: Dict[str, Any], room: Room, locations
...
+    text += "<STYLE>body {font-family: Courier;}</STYLE></HEAD><BODY><table><tr>"
+    for cardnum in range(0, len(slot_data["cards"])):
+        text += "<td>|--------------|<BR>"
+        text += "|B&nbsp;&nbsp;I&nbsp;&nbsp;N&nbsp;&nbsp;G&nbsp;&nbsp;O&nbsp;|<br>"
```
Comment (black-sliver) [inline WebHostLib/tracker.py:915]: probably better use css letter-spacing or pre tag ("preformatted text") depending on the style

File: worlds/bingo/Items.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    progression: bool = True
+    event: bool = False
+
+
+class BingoItem(Item):
+    game: str = "Bingo"
+
+
+item_table = {
+    "Bingo Call 01": ItemData(900),
```
Comment (black-sliver) [inline worlds/bingo/Items.py:16]: this could be reduced with a range() and string formatting with simple math

File: worlds/bingo/__init__.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    def create_regions(self):
+        create_regions(self.world, self.player)
+
+    def generate_output(self, output_directory: str):
+        filename = f"AP-{self.world.seed_name}-P{self.player}-{self.world.player_name[self.player]}.html"
+        out_file = os.path.join(output_directory, filename)
+        with open(out_file, 'w') as f:
+            # Web design is my passion
+            f.write("<HTML><HEAD><STYLE>body {font-family: Courier;}</STYLE></HEAD><BODY>")
```
Comment (black-sliver) [inline worlds/bingo/__init__.py:97]: see above

File: WebHostLib/tracker.py (hunk 2 above)
Comment (Alchav) [reply WebHostLib/tracker.py:915]: This is more of a proof of concept. But I am very limited in my knowledge/experience with CSS and web design generally. I would appreciate any help I could get on this

**Comment by JarnoWesthof** [comment]
I like this change where specific games can run additional code on the server if they deem neceserly, the move to the world class is a great choice for that

**Pull request closed without merging**